File: examples/csharp/avalonia/NotesShowcase/ViewModels/NotebooksRootVM.cs
using System.Collections.ObjectModel;
using System.Reactive.Concurrency;
using System.Windows.Input;
using VMx.Builders;
using VMx.Capabilities;
using VMx.Commands;
using VMx.Components;
using VMx.Hierarchical;
using VMx.Messages;
using VMx.Notifications;
using VMx.Services;
using NotesShowcase.Models;

namespace NotesShowcase.ViewModels;

/// <summary>
/// Root of the notebooks tree.
///
/// VMx-API adaptation (plan §3.a.7): the plan asked for
/// <c>HierarchicalVM&lt;NotebookModel, NotebookVM&gt;</c>, but
/// <see cref="HierarchicalVM{TModel,TVM}"/> materializes its children from a
/// factory at construct time and is awkward to mutate dynamically (it has
/// <c>AddChild</c>, but no first-class "current selection" / "walk" surface
/// the plan assumed). Instead we own a flat list of <see cref="NotebookVM"/>
/// instances, exposing <see cref="Roots"/> and <see cref="Walk"/> for the
/// view; <see cref="AddNotebookAsync"/> persists via the repo and re-publishes
/// a <see cref="TreeStructureChangedMessage"/> on the hub so subscribers
/// observe structural changes the same way they would on
/// <see cref="HierarchicalVM{TModel,TVM}"/>.
/// </summary>
public sealed class NotebooksRootVM
    : ComponentVMBase,
      INewCreatable,
      IReconstructable
{
    private readonly INoteRepository _repo;
    private readonly IMessageHub _hub;
    private readonly IDispatcher _dispatcher;
    private readonly INotificationHub? _notificationHub;
    private readonly ObservableCollection<NotebookVM> _all = new();
    private NotebookVM? _current;

    /// <inheritdoc/>
    public override ViewModelType Type => ViewModelType.Component;

    /// <summary>Public hub accessor.</summary>
    public new IMessageHub Hub => base.Hub;

    /// <summary>All notebooks (flat, ordered).</summary>
    public ObservableCollection<NotebookVM> All => _all;

    /// <summary>Root notebooks (no parent).</summary>
    public IReadOnlyList<NotebookVM> Roots
        => _all.Where(nb => nb.Model.ParentId is null).ToList();

    /// <summary>Iterates every notebook in the tree (parents-before-children, repo order).</summary>
    public IEnumerable<NotebookVM> Walk() => _all;

    /// <summary>Children of <paramref name="parent"/> (by ParentId).</summary>
    public IReadOnlyList<NotebookVM> ChildrenOf(NotebookVM parent)
        => _all.Where(nb => nb.Model.ParentId == parent.Model.Id).ToList();

    /// <summary>Currently selected notebook (two-way bindable).</summary>
    public NotebookVM? Current
    {
        get => _current;
        set
        {
            if (ReferenceEquals(_current, value)) return;
            _current = value;
            Hub.Send(PropertyChangedMessage<IComponentVM>.Create(this, Name, nameof(Current)));
            RaisePropertyChanged(nameof(Current));
        }
    }

    // ── INewCreatable ──────────────────────────────────────────────────────

    /// <inheritdoc/>
    public bool CanCreateNew() => Status == VMx.Lifecycle.ConstructionStatus.Constructed;

    /// <inheritdoc/>
    public void CreateNew()
    {
        // Synchronous capability — schedules a default "New Notebook" via fire-and-forget.
        _ = AddNotebookAsync(parentId: null, name: "New Notebook");
    }

    /// <summary>
    /// View-facing command parity with the plan: bound to a toolbar button.
    /// Same default behavior as <see cref="CreateNew"/>.
    /// </summary>
    public ICommand AddNotebookCommand { get; }

    /// <summary>
    /// Persists a new notebook via the repository and re-publishes a
    /// <see cref="TreeStructureChangedMessage"/> so subscribers refresh.
    /// </summary>
    public async Task AddNotebookAsync(string? parentId, string name, CancellationToken ct = default)
    {
        // "nb-" + 5 random hex chars (8 chars total). Same suffix length as
        // notes ("note-" + 5 chars = 10 total) — only the prefix differs.
        var id = $"nb-{Guid.NewGuid():N}"[..8];
        var model = new NotebookModel(id, name, parentId);
        await _repo.AddNotebookAsync(model, ct).ConfigureAwait(false);
        var vm = NotebookVM.Builder()
            .Name($"nb:{id}")
            .Model(model)
            .Services(_hub, _dispatcher)
            .ChildrenGetter(ChildrenOf)
            .Build();
        vm.Construct();
        _all.Add(vm);
        // Phase 5.a binding gap #2: tell the new VM's parent (if any) to
        // refresh its Children view so an already-bound TreeView picks up
        // the new nested node.
        if (parentId is not null)
        {
            var parent = _all.FirstOrDefault(nb => nb.Model.Id == parentId);
            parent?.NotifyChildrenChanged();
        }
        else
        {
            // Root-level add: Roots is a computed snapshot, so an
            // already-bound TreeView only re-reads it on an explicit raise.
            // Marshal to the foreground — this continuation runs off the UI
            // thread after ConfigureAwait(false).
            _dispatcher.Foreground.Schedule(() => RaisePropertyChanged(nameof(Roots)));
        }
        // Index is the new tail of _all (we just appended).
        Hub.Send(new TreeStructureChangedMessage(
            Source: this,
            Change: TreeStructureChange.Added,
            Affected: vm,
            Index: _all.Count - 1));
        // Spec §6.2: publish a "Notebook added" notification so the toast
        // region surfaces it. No-op when no hub is wired.
        if (_notificationHub is not null)
        {
            _ = _notificationHub.Post(new Notification(
                NotificationType.Notification,
                $"Notebook added: “{name}”"));
        }
    }

    /// <summary>
    /// Loads notebooks from the repository, replaces the internal collection,
    /// and constructs each child VM. Called by <see cref="WorkspaceVM"/> during
    /// async OnConstruct (plan §3.a.11).
    /// </summary>
    public async Task PopulateAsync(CancellationToken ct = default)
    {
        var (notebooks, _) = await _repo.LoadAllAsync(ct).ConfigureAwait(false);

        // Dispose previous children before replacing.
        foreach (var prev in _all) prev.Dispose();
        _all.Clear();
        _current = null;

        foreach (var nb in notebooks)
        {
            var vm = NotebookVM.Builder()
                .Name($"nb:{nb.Id}")
                .Model(nb)
                .Services(_hub, _dispatcher)
                .ChildrenGetter(ChildrenOf)
                .Build();
            vm.Construct();
            _all.Add(vm);
        }

        // Reset notification — subscribers refresh their tree projections.
        Hub.Send(new TreeStructureChangedMessage(
            Source: this,
            Change: TreeStructureChange.Added,
            Affected: this,
            Index: -1));

        // Roots is a computed snapshot: a TreeView bound before this populate
        // completed (the App sets DataContext without awaiting ConstructAsync)
        // only re-reads it on an explicit raise. Marshal to the foreground —
        // this continuation runs off the UI thread after ConfigureAwait(false).
        _dispatcher.Foreground.Schedule(() => RaisePropertyChanged(nameof(Roots)));
    }

    private NotebooksRootVM(
        string name,
        string hint,
        IMessageHub hub,
        IDispatcher dispatcher,
        INoteRepository repo,
        INotificationHub? notificationHub)
        : base(name, hint, hub, dispatcher, onConstruct: null, onDestruct: null)
    {
        _repo = repo;
        _hub = hub;
        _dispatcher = dispatcher;
        _notificationHub = notificationHub;

        AddNotebookCommand = RelayCommand.Builder()
            .Predicate(CanCreateNew)
            .Task(CreateNew)
            .Build();
    }

    /// <inheritdoc/>
    protected override void OnDestruct()
    {
        foreach (var nb in _all) nb.Destruct();
        base.OnDestruct();
    }

    /// <inheritdoc/>
    protected override void OnDispose()
    {
        foreach (var nb in _all) nb.Dispose();
        (AddNotebookCommand as IDisposable)?.Dispose();
        base.OnDispose();
    }

    /// <summary>Returns a new empty builder.</summary>
    public static NotebooksRootVMBuilder Builder() => NotebooksRootVMBuilder.Empty;

    /// <summary>Immutable fluent builder.</summary>
    public sealed class NotebooksRootVMBuilder
    {
        private readonly string? _name;
        private readonly string _hint;
        private readonly IMessageHub? _hub;
        private readonly IDispatcher? _dispatcher;
        private readonly INoteRepository? _repo;
        private readonly INotificationHub? _notificationHub;

        internal static readonly NotebooksRootVMBuilder Empty = new();
        private NotebooksRootVMBuilder() { _hint = ""; }
        private NotebooksRootVMBuilder(
            string? name, string hint,
            IMessageHub? hub, IDispatcher? dispatcher, INoteRepository? repo,
            INotificationHub? notificationHub)
        {
            _name = name; _hint = hint; _hub = hub; _dispatcher = dispatcher;
            _repo = repo; _notificationHub = notificationHub;
        }

        /// <summary>Sets the required Name.</summary>
        public NotebooksRootVMBuilder Name(string name) => new(name, _hint, _hub, _dispatcher, _repo, _notificationHub);
        /// <summary>Sets the optional Hint.</summary>
        public NotebooksRootVMBuilder Hint(string hint) => new(_name, hint, _hub, _dispatcher, _repo, _notificationHub);
        /// <summary>Sets the required Services.</summary>
        public NotebooksRootVMBuilder Services(IMessageHub hub, IDispatcher dispatcher)
            => new(_name, _hint, hub, dispatcher, _repo, _notificationHub);
        /// <summary>Sets the required repository.</summary>
        public NotebooksRootVMBuilder Repository(INoteRepository repo)
            => new(_name, _hint, _hub, _dispatcher, repo, _notificationHub);
        /// <summary>
        /// Sets the optional notification hub. When set,
        /// <see cref="AddNotebookAsync"/> publishes a "Notebook added"
        /// notification (spec §6.2).
        /// </summary>
        public NotebooksRootVMBuilder NotificationHub(INotificationHub notificationHub)
            => new(_name, _hint, _hub, _dispatcher, _repo, notificationHub);

        /// <summary>Builds the VM after validation.</summary>
        public NotebooksRootVM Build()
        {
            BuilderValidationException.Require(_name, "Name");
            BuilderValidationException.Require(_hub, "Hub");
            BuilderValidationException.Require(_dispatcher, "Dispatcher");
            BuilderValidationException.Require(_repo, "Repository");
            return new NotebooksRootVM(_name!, _hint, _hub!, _dispatcher!, _repo!, _notificationHub);
        }
    }
}

File: examples/typescript/react/notes-showcase/src/viewmodels/notebooksRootVM.ts
/**
 * NotebooksRootVM — root of the notebooks tree.
 *
 * VMx-API adaptation (parity with C# / Python flavors): the plan asked for
 * `HierarchicalVM<NotebookModel, NotebookVM>`, but VMx TS's
 * `HierarchicalVM` sources each node's children from a per-node factory
 * (materialized lazily on first access, not eagerly at construct time) and
 * lacks the "current selection" / "walk" surface the showcase needs.
 * We own a flat collection and expose roots / children-of / walk over it.
 * `addNotebookAsync` persists via the repo and publishes a
 * `TreeStructureChangedMessage` so subscribers observe structural changes
 * the same way they would on a `HierarchicalVM`.
 */
import {
  ComponentVMBase,
  ConstructionStatus,
  declareCapabilities,
  RelayCommand,
  TreeStructureChangedMessage,
  ViewModelType,
  type ICommand,
  type IDispatcher,
  type IMessageHub,
} from "@thekaveh/vmx";
import {
  type INotificationHub,
  Notification,
  NotificationType,
} from "@thekaveh/vmx/notifications";

import type { NotebookModel } from "../models/notebookModel.js";
import type { INoteRepository } from "../models/noteRepository.js";
import { NotebookVM } from "./notebookVM.js";

const SENTINEL = Symbol("not-set");

let counter = 0;
function newNotebookId(): string {
  counter += 1;
  return `nb-new-${Date.now().toString(36)}-${counter}`;
}

export class NotebooksRootVM extends ComponentVMBase {
  readonly #repo: INoteRepository;
  readonly #notificationHub: INotificationHub | null;
  readonly #all: NotebookVM[] = [];
  readonly #addNotebookCommand: RelayCommand;
  #current: NotebookVM | null = null;

  constructor(opts: {
    name: string;
    hint: string;
    hub: IMessageHub;
    dispatcher: IDispatcher;
    repository: INoteRepository;
    notificationHub?: INotificationHub | null;
  }) {
    super({
      name: opts.name,
      hint: opts.hint,
      hub: opts.hub,
      dispatcher: opts.dispatcher,
    });
    this.#repo = opts.repository;
    this.#notificationHub = opts.notificationHub ?? null;
    declareCapabilities(this, "INewCreatable", "IReconstructable");

    this.#addNotebookCommand = RelayCommand.builder()
      .predicate(() => this.canCreateNew())
      .task(() => {
        void this.addNotebookAsync(null, "New Notebook");
      })
      .build();
  }

  get type(): ViewModelType {
    return ViewModelType.Component;
  }

  get hub(): IMessageHub {
    return this._hub;
  }

  /** All notebooks (flat, ordered). */
  get all(): readonly NotebookVM[] {
    return this.#all;
  }

  /** Root notebooks (no parent). */
  get roots(): NotebookVM[] {
    return this.#all.filter((nb) => nb.model.parentId === null);
  }

  /** Iterates every notebook in repository order. */
  walk(): Iterable<NotebookVM> {
    return this.#all.slice();
  }

  /** Returns the children of the given notebook (matched by parentId). */
  childrenOf(parent: NotebookVM): NotebookVM[] {
    return this.#all.filter((nb) => nb.model.parentId === parent.model.id);
  }

  /** Currently selected notebook (two-way bindable). */
  get current(): NotebookVM | null {
    return this.#current;
  }

  set current(value: NotebookVM | null) {
    if (this.#current === value) return;
    this.#current = value;
    this._notifyPropertyChanged("current");
  }

  // ── INewCreatable ─────────────────────────────────────────────────────────

  canCreateNew(): boolean {
    return this.status === ConstructionStatus.Constructed;
  }

  createNew(): void {
    void this.addNotebookAsync(null, "New Notebook");
  }

  get addNotebookCommand(): ICommand {
    return this.#addNotebookCommand;
  }

  /**
   * Persists a new notebook via the repo and publishes a
   * `TreeStructureChangedMessage("added")` so subscribers refresh.
   */
  async addNotebookAsync(parentId: string | null, name: string): Promise<void> {
    const model: NotebookModel = { id: newNotebookId(), name, parentId };
    await this.#repo.addNotebook(model);
    const vm = NotebookVM.builder()
      .name(`nb:${model.id}`)
      .model(model)
      .services(this._hub, this._dispatcher)
      .build();
    vm.construct();
    this.#all.push(vm);
    // Real-wiring audit, pass 6: useVm re-renders only on
    // PropertyChangedMessage — TreeStructureChangedMessage alone left the
    // new notebook invisible until the next unrelated re-render.
    this._notifyPropertyChanged("roots");
    this._hub.send(
      new TreeStructureChangedMessage(
        this,
        this._name,
        "added",
        vm,
        this.#all.length - 1,
      ),
    );
    // Spec §6.2: publish a "Notebook added" notification (cross-flavor parity
    // with the C# and Python flavors). No-op when no notification hub is
    // wired (e.g. in unit tests that don't care about toast feedback).
    if (this.#notificationHub !== null) {
      void this.#notificationHub.post(
        new Notification(NotificationType.Notification, `Notebook added: “${name}”`),
      );
    }
  }

  /**
   * Loads notebooks from the repository and constructs each child VM.
   * Called by `WorkspaceVM` during async construct.
   */
  async populateAsync(): Promise<void> {
    const { notebooks } = await this.#repo.loadAll();
    for (const prev of this.#all) prev.dispose();
    this.#all.length = 0;
    this.#current = null;

    for (const nb of notebooks) {
      const vm = NotebookVM.builder()
        .name(`nb:${nb.id}`)
        .model(nb)
        .services(this._hub, this._dispatcher)
        .build();
      vm.construct();
      this.#all.push(vm);
    }

    this._hub.send(
      new TreeStructureChangedMessage(this, this._name, "added", this, -1),
    );
  }

  protected override _onDestruct(): void {
    for (const nb of this.#all) nb.destruct();
    super._onDestruct();
  }

  protected override _onDispose(): void {
    for (const nb of this.#all) nb.dispose();
    this.#addNotebookCommand.dispose();
    super._onDispose();
  }

  static builder(): NotebooksRootVMBuilder {
    return new NotebooksRootVMBuilder();
  }
}

export class NotebooksRootVMBuilder {
  #name: string | null = null;
  #hint = "";
  #hub: IMessageHub | null = null;
  #dispatcher: IDispatcher | null = null;
  #repo: INoteRepository | typeof SENTINEL = SENTINEL;
  #notificationHub: INotificationHub | null = null;

  constructor(from?: NotebooksRootVMBuilder) {
    if (from) {
      this.#name = from.#name;
      this.#hint = from.#hint;
      this.#hub = from.#hub;
      this.#dispatcher = from.#dispatcher;
      this.#repo = from.#repo;
      this.#notificationHub = from.#notificationHub;
    }
  }

  name(value: string): NotebooksRootVMBuilder {
    const b = new NotebooksRootVMBuilder(this);
    b.#name = value;
    return b;
  }

  hint(value: string): NotebooksRootVMBuilder {
    const b = new NotebooksRootVMBuilder(this);
    b.#hint = value;
    return b;
  }

  services(hub: IMessageHub, dispatcher: IDispatcher): NotebooksRootVMBuilder {
    const b = new NotebooksRootVMBuilder(this);
    b.#hub = hub;
    b.#dispatcher = dispatcher;
    return b;
  }

  repository(repo: INoteRepository): NotebooksRootVMBuilder {
    const b = new NotebooksRootVMBuilder(this);
    b.#repo = repo;
    return b;
  }

  notificationHub(hub: INotificationHub): NotebooksRootVMBuilder {
    const b = new NotebooksRootVMBuilder(this);
    b.#notificationHub = hub;
    return b;
  }

  build(): NotebooksRootVM {
    if (this.#name === null) throw new Error("name is required");
    if (this.#hub === null || this.#dispatcher === null)
      throw new Error("services (hub + dispatcher) are required");
    if (this.#repo === SENTINEL) throw new Error("repository is required");
    return new NotebooksRootVM({
      name: this.#name,
      hint: this.#hint,
      hub: this.#hub,
      dispatcher: this.#dispatcher,
      repository: this.#repo,
      notificationHub: this.#notificationHub,
    });
  }
}
